README.md
# Yandex Backend School Task
[![version][version-badge]][CHANGELOG]

REST API сервис для обработки выгрузок данных интернет-магазина подарков.

## Содержание
>1\. [Описание сервиса](https://github.com/dmitriev-z/backend_yandex#1-описание-сервиса)
>>1.1 [Обработчики запросов](https://github.com/dmitriev-z/backend_yandex#11-обработчики-запросов)    
>>>1.1.1 [POST /imports](https://github.com/dmitriev-z/backend_yandex#111-post-imports)  
>>>1.1.2 [PATCH /imports/$import_id/citizens/$citizen_id](https://github.com/dmitriev-z/backend_yandex#112-patch-importsimport_idcitizenscitizen_id)   
>>>1.1.3 [GET /imports/$import_id/citizens](https://github.com/dmitriev-z/backend_yandex#113-get-importsimport_idcitizens)  
>>>1.1.4 [GET /imports/$import_id/citizens/birthdays](https://github.com/dmitriev-z/backend_yandex#114-get-importsimport_idcitizensbirthdays)  
>>>1.1.5 [GET /imports/$import_id/towns/stat/percentile/age](https://github.com/dmitriev-z/backend_yandex#115-get-importsimport_idtownsstatpercentileage)
>>
>>1.2 [Управление сервисом](https://github.com/dmitriev-z/backend_yandex#12-управление-сервисом)
>
>2\. [Установка сервиса](https://github.com/dmitriev-z/backend_yandex#2-установка-сервиса)
>>2.1 [Установка требуемых пакетов](https://github.com/dmitriev-z/backend_yandex#21-установка-требуемых-пакетов)  
>>2.2 [Установка MongoDB Community Edition](https://github.com/dmitriev-z/backend_yandex#22-установка-mongodb-community-edition)  
>>2.3 [Создание виртуального окружения](https://github.com/dmitriev-z/backend_yandex#23-создание-виртуального-окружения)  
>>2.4 [Установка требуемых python библиотек](https://github.com/dmitriev-z/backend_yandex#24-установка-требуемых-python-библиотек)
>
>3\. [Запуск сервиса](https://github.com/dmitriev-z/backend_yandex#3-запуск-сервиса)
>>3.1. [Запуск сервиса локально](https://github.com/dmitriev-z/backend_yandex#31-запуск-сервиса-локально)  
>>3.2. [Запуск на удаленном сервере](https://github.com/dmitriev-z/backend_yandex#32-запуск-на-удаленном-сервере)
>
>4\. [Тестирование сервиса](https://github.com/dmitriev-z/backend_yandex#4-тестирование-сервиса)

## 1. Описание сервиса
Сервис написан на `Python 3.6` и запущен в виртуальной окружении с именем `env`.    
Сервис развернут на виртуальной машине `84.201.156.229` на порту `8080` .  
В качестве хранилища данных используется NoSQL база данных MongoDB Community Edition. 
База данных развернута на той же виртуальной машине на порту `27017`.

### 1.1. Обработчики запросов
Сервис поддерживает 3 обработчика запросов:

|**№**|**URL**| **Method**| 
|:---:|---|:---:| 
|1|`http://84.201.156.229/imports`|*POST*|
|2|`http://84.201.156.229/imports/$import_id/citizens/$citizen_id`|*PATCH*|
|3|`http://84.201.156.229/imports/$import_id/citizens`|*GET*|
|4|`http://84.201.156.229/imports/$import_id/citizens/birthdays` |*GET*|
|5|`http://84.201.156.229/imports/$import_id/towns/stat/percentile/age`|*GET*|

#### 1.1.1 POST /imports
Принимает на вход набор с данными о жителях в формате json и сохраняет его с уникальным идентификатором import_id.  
В наборе данных для каждого жителя должны присутствовать все поля, значения не могут быть null , порядок полей не важен:

|**Поле** | **Тип** | **Значение** |
|:---:|:---:|---|
|citizen_id| целое число| Уникальный идентификатор жителя, неотрицательное число.|
|town| строка| Название города. Непустая строка, содержащая хотя бы 1 букву или цифру, не более 256 символов.|
|street| строка| Название улицы. Непустая строка, содержащая хотя бы 1 букву или цифру, не более 256 символов.|
|building| строка| Номер дома, корпус и строение. Непустая строка,содержащая хотя бы 1 букву или цифру, не более 256 символов.|
|apartment| целое число| Номер квартиры, неотрицательное число.|
|name| строка| Непустая строка, не более 256 символов.|
|birth_date| строка| Дата рождения в формате ДД.ММ.ГГГГ (UTC +0). Должна быть меньше текущей даты.|
|gender| строка| Значения male , female.|
|relatives| список из целых чисел| Ближайшие родственники, уникальные значения существующих citizen_id жителей из этой же выгрузки.|

Пример запроса:
```
POST /imports
{
    "citizens": [
        {
            "citizen_id": 1,
            "town": "Москва",
            "street": "Льва Толстого",
            "building": "16к7стр5",
            "apartment": 7,
            "name": "Иванов Иван Иванович",
            "birth_date": "26.12.1986",
            "gender": "male",
            "relatives": [2]
        },
        {
            "citizen_id": 2,
            "town": "Москва",
            "street": "Льва Толстого",
            "building": "16к7стр5",
            "apartment": 7,
            "name": "Иванов Сергей Иванович",
            "birth_date": "01.04.1997",
            "gender": "male",
            "relatives": [1]
        },
        {
            "citizen_id": 3,
            "town": "Керчь",
            "street": "Иосифа Бродского",
            "building": "2",
            "apartment": 11,
            "name": "Романова Мария Леонидовна",
            "birth_date": "23.11.1986",
            "gender": "female",
            "relatives": []
        }
    ]
}
```
В случае успешного добавления данных сервис вернет ответ с HTTP статусом `201 Created` и идентификатором импорта:
```
HTTP 201
{
    "data": {
        "import_id": 1
    }
}
```
В случае, если входные данные не прошли валидацию 
(данные не были переданы, отсутствуют обязательные поля, присутсвуют неописанные поля, неправильное значение поля), 
сервис вернет ответ `400: Bad Request`.

#### 1.1.2. PATCH /imports/$import_id/citizens/$citizen_id
Изменяет информацию о жителе `$citizen_id` в указанном наборе данных `$import_id`.
На вход подается JSON в котором можно указать любые данные о жителе, кроме `citizen_id` : `name` , `gender` , `birth_date (UTC)`, `relatives` , `town` , `street` , `building` , `apartment` .  
В запросе должно быть указано хотя бы одно поле, значения не могут быть null.

Пример запроса:
```
PATCH /imports/1/citizens/3
{
    "name": "Иванова Мария Леонидовна",
    "town": "Москва",
    "street": "Льва Толстого",
    "building": "16к7стр5",
    "apartment": 7,
    "relatives": [1]
}
```
В случае успешного обновления данных сервис вернет ответ с HTTP статусом `200 OK` и актуальную информацию об указанном жителе:
```
HTTP 200
{
    "data": {
        "citizen_id": 3,
        "town": "Москва",
        "street": "Льва Толстого",
        "building": "16к7стр5",
        "apartment": 7,
        "name": "Иванова Мария Леонидовна",
        "birth_date": "23.11.1986",
        "gender": "female",
        "relatives": [1]
    }
}
```
В случае, если входные данные не прошли валидацию (данные не были переданы, присутсвуют неописанные поля, неправильное значение поля), сервис вернет ответ `400: Bad Request`.

#### 1.1.3. GET /imports/$import_id/citizens
Возвращает список всех жителей для указанного набора данных. 
 
Если набор данных `$import_id` присутствует в базе данных, то сервис вернет список всех citizens этого набора данных:
```
HTTP 200
{
    "data": [
        {
            "citizen_id": 1,
            "town": "Москва",
            "street": "Льва Толстого",
            "building": "16к7стр5",
            "apartment": 7,
            "name": "Иванов Иван Иванович",
            "birth_date": "26.12.1986",
            "gender": "male",
            "relatives": [2,3]
        },
        {
            "citizen_id": 2,
            "town": "Москва",
            "street": "Льва Толстого",
            "building": "16к7стр5",
            "apartment": 7,
            "name": "Иванов Сергей Иванович",
            "birth_date": "01.04.1997",
            "gender": "male",
            "relatives": [1]
        },
        {
            "citizen_id": 3,
            "town": "Москва",
            "street": "Льва Толстого",
            "building": "16к7стр5",
            "apartment": 7,
            "name": "Иванова Мария Леонидовна",
            "birth_date": "23.11.1986",
            "gender": "female",
            "relatives": [1]
        }
    ]
}
```
Если набор данных `$import_id` отсутсвует в базе данных, то сервис вернет ответ `400: Bad Request`.

#### 1.1.4. GET /imports/$import_id/citizens/birthdays
Возвращает жителей и количество подарков, которые они будут покупать своим ближайшим родственникам (1-го порядка), 
сгруппированных по месяцам из указанного набора данных.  
Если в импорте в каком-либо месяце нет ни одного жителя с днями рождения
ближайших родственников, значением такого ключа должен быть пустой список.  

Если набор данных `$import_id` присутствует в базе данных, сервис вернет список жителей и количество подарков, 
которые они будут покупать своим ближайшим родственникам, сгруппированных по месяцам из указанного набора данных:
```
HTTP 200
{
    "data": {
        "1": [],
        "2": [],
        "3": [],
        "4": [{
            "citizen_id": 1,
            "presents": 1,
        }],
        "5": [],
        "6": [],
        "7": [],
        "8": [],
        "9": [],
        "10": [],
        "11": [{
            "citizen_id": 1,
            "presents": 1
        }],
        "12": [
            {
                "citizen_id": 2,
                "presents": 1
            },
            {
                "citizen_id": 3,
                "presents": 1
            }
        ]
    }
}
```
Если набор данных `$import_id` отсутсвует в базе данных, то сервис вернет ответ `400: Bad Request`.

#### 1.1.5. GET /imports/$import_id/towns/stat/percentile/age
Возвращает статистику по городам для указанного набора данных `$import_id` в разрезе возраста (полных лет) жителей: 
`p50`, `p75`, `p99`, где число - это значение перцентиля.  
Если набор данных `$import_id` присутствует в базе данных, 
сервис вернет статистику по городам в разрезе возраста (полных лет) жителей:
```
HTTP 200
{
    "data": [
        {
            "town": "Москва",
            "p50": 35.0,
            "p75": 47.5,
            "p99": 59.5
        },
        {
            "town": "Санкт-Петербург",
            "p50": 45.0,
            "p75": 52.5,
            "p99": 97.15
        }
    ]
}
```
Если набор данных `$import_id` отсутсвует в базе данных, то сервис вернет ответ `400: Bad Request`.

### 1.2. Управление сервисом
Для управления сервисом на виртуальной машине был создан системный сервис `yandexbackend`.  
Конфигурационный файл сервиса доступен по пути `/etc/systemd/system/yandexbackend.service`.

Для остановки сервиса выполните в терминале следующую команду:
```shell script
sudo systemctl stop yandexbackend
```
Для запуска сервиса выполните в терминале следующую команду:
```shell script
sudo systemctl stop yandexbackend
```
Для перезагрузки сервиса выполните в терминале следующую команду:
```shell script
sudo systemctl restart yandexbackend
```
Для просмотра состояния сервиса выполните в терминале следующую команду:
```shell script
sudo systemctl status yandexbackend
```

## 2. Установка сервиса
Для установки сервиса в первую очередь необходимо склонировать данный репозиторий.  
После этого перейдите в папку с склонироованным репозиторием.

### 2.1. Установка требуемых пакетов
Для работы с python библиотеками необходимо установить менеджер пакетов [pip](https://pip.pypa.io/en/stable/).  
Для установки `pip` выполните в терминале следующие команды:
```shell script
sudo apt-get update
sudo apt-get install python3-pip
```

### 2.2 Установка MongoDB Community Edition
Подробная инструкция по установке MongoDB Community Edition представлена на [официальном сайте](https://docs.mongodb.com/v3.2/administration/install-community/).  
После установки базы данных необхродимо включить её автозапуск после перезагрузки машины.
Для этого выполните в терминале следующую команду:
```shell script
sudo systemctl enable mongod
```

### 2.3. Создание виртуального окружения
Для корректной работы сервиса необходимо изолировать его от других python-приложений. Для этого необходимо создать виртуальное окружение.  
Для работы с виртуальными окружениями необходимо установить библиотеку [virtualenv](https://virtualenv.pypa.io/en/latest/)
Для установки `virtualenv` выполните в терминале следующую команду:
```shell script
sudo pip3 install virtualenv
```
Далее необходимо создать виртуальную окружение. Для этого выполните в терминале следующую команду:
```shell script
virtualenv {env_name}
```
Данная команда создаст новое виртуальное окружение с именем `{env_name}` в катологе `{env_name}` текущей директории.

Далее необходимо включить созданное виртуальное окружение. Для этого выполните в терминале следующую команду:
```shell script
source {env_name}/bin/activate
```

### 2.4. Установка требуемых python библиотек
Для работы сервиса необходима установка следующих python-библиотек:
- [Flask](https://palletsprojects.com/p/flask/)  
Flask - фреймворк WSGI веб-приложений.
- [Gunicorn](https://gunicorn.org/)  
Gunicorn - WSGI HTTP сервер.
- [Cerberus](http://docs.python-cerberus.org/en/stable/)  
Cerberus - библиотека, предоставляющая инструменты для валидации данных.
- [Pymongo](https://api.mongodb.com/python/current/index.html)  
Pymongo - библитотека, предоставляющая инструменты для работы с MongoDB.
- [Pytest](https://docs.pytest.org/en/latest/)  
Pytest - библиоткека для написания тестов. 
- [Requests](https://2.python-requests.org/en/master/)  
Requests - библиотека для работы с HTTP забросами. В данном проекте используется в тестах.
- [Numpy](https://numpy.org/)  
Numpy - библиотека для научных вычислений.

Все требуемые библиотеки описаны в файле `requirements.txt`.  
Для установки требуемых библиотек выполните в терминале следующую команду:
```shell script
sudo pip3 install -r requirements.txt
```

## 3. Запуск сервиса

### 3.1. Запуск сервиса локально
Для запуска сервиса локально необходимо указать фрейворку Flask с каким приложением работать.  
Для этого, находясь в корневом каталоге сервиса, выполните в терминале следующую команду:
```shell script
export FLASK_APP=service.service_api.py
```
После этого можно запустить сервис, выполнив в терминале следующую команду:
```shell script
flask run
```
Данная команда запустит сервис на локальной машине на порту `5000`.

Если вы хотите запустить сервис на другом порту, то задайте аргумент `--port` при выполнениее предыдущей команды:
```shell script
flask run --port={port_to_run_app}
```
Таким образом приложение будет запущено на порту `{port_to_run_app}`

### 3.2. Запуск на удаленном сервере

Для удобства работы с сервисом на удаленном сервере необходимо создать соответсвующий системный сервис.  
Конечно, можно было бы запустить сервис вручную, но системный сервис  позволит автоматически запускать Gunicorn и обслуживать приложение Flask через систему инициализации.

Для создания системного сервиса необходимо создать файл с расширением `.service` в каталоге `/etc/systemd/system`.  
Для этого выполните в терминале следующую команду:
```shell script
sudo touch /etc/systemd/system/{service_name}.service
```
Где `{service_name}` - имя создаваемого системного сервиса.

Далее с помощью текстового редактора (nano, vim) необходимо добавить в созданный файл следующую информацию:
```text
[Unit]
Description=Gunicorn instance to serve service
After=network.target
[Service]
User={user}
Group=www-data
WorkingDirectory={service_directory}
Environment="PATH={service_directory}/env/bin"
ExecStart={service_directory}/env/bin/gunicorn --workers 1 --bind 0.0.0.0:{service_port} service.wsgi:app
[Install]
WantedBy=multi-user.target
```
Раздел `[Unit]` определяет метаданные и зависимости приложения. 
 
В разделе `[Service\` указываются:
1. Пользователь `User={user}` и группа `Group=www-data`, с помощью которых будет запущен сервис.  
2. Рабочий каталог сервиса `WorkingDirectory={service_directory}`, а так же путь к виртуальному окружению `Environment="PATH={service_directory}/env/bin"`.
3. Команда, с помощью которйо будет запускаться сервис `ExecStart={service_directory}/env/bin/gunicorn --workers 1 --bind 0.0.0.0:{service_port} service.wsgi:app`  
Согласно данной команде, система будет запускать один рабочий процесс на порту `{service_port}`.

Раздел `[Install]` определяет, к чему должен подключиться сервис во время автозапуска.

Закройте файл, сохранив изменения.

Далее необходимо запустить сервис, выполнив в терминале следующую команду:
```shell script
sudo systemctl start {service_name}
```
А так же включить сервис в автозапуск, выполнив в терминале слледующую команду:
```shell script
sudo systemctl enable {service_name}
```

## 4. Тестирование сервиса
|**[ВАЖНО\]**| **Не запускайте тесты на машине, где запущена рабочая версия сервиса!**| **[ВАЖНО\]**|
|:---:|:---:|:---:|
|**[ВАЖНО\]**| **Тесты ожидают, что и они и сервис будут работать с чистой базой данных.** |**[ВАЖНО\]**|
|**[ВАЖНО\]**| **По завершении тесты удаляют все таблицы, созданные в базе данных.**| **[ВАЖНО\]**|
|**[ВАЖНО\]**| **Запуск тестов на машине, где запущена рабочая версия сервиса может привести к сбоям в его работе.**| **[ВАЖНО\]**|
---
Для тестирования необходимо запустить сервис локально, как это описано в разделе [3.1](https://github.com/dmitriev-z/backend_yandex#31-запуск-сервиса-локально).

---
Перед тестированием необходимо убедиться, что базы данных `yandexbackend` не существует.  
Для этого необходимо перейти в shell MongoDB, выполнив в терминале команду:
```shell script
mongo
```
Откроется shell MongoDB. В нем выполните команду:
```shell script
show dbs
```
Если в списке баз данных присутствует `yandexbackend`, то необходимо удалить эту базу данных.  
Если база данных `yandexbackend` отсутствует в списке существующих баз данных, то можно закрыть shell MongoBB, выполнив команду:
```shell script
quit()
```
Для удаления базы данных `yandexbackend` выполните следующие команды:
```shell script
use yandexbackend
db.dropDatabase()
```
---

Для запуска тестов из корневой директории сервиса выполните в терминале следующую команду:
```shell script
pytest tests/test_service.py -v
```
Для запуска тестов из директории `/tests` выполните в терминале следующую команду:
```shell script
pytest test_service.py -v
```
Для тестов API используется баблиотека `requests`.  
По умолчанию тесты будут отправлять запросы по адресу `http://127.0.0.1:5000/{path}`.
Если вы запустили сервис на порту, отличном от `5000`, то для этого предусмотрен аргумент `--service-address`.  
Чтобы тесты отправляли запросы по другому адресу, запустите тесты, выполнив в терминале следующую команду:
```shell script
pytest test_service.py --service-address 127.0.0.1:{port}
```
Где `{port}` - порт, на котором запущен сервис.


[CHANGELOG]: ./CHANGELOG.md
[version-badge]: https://img.shields.io/badge/version-4.0.0-blue.svg


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

## [4.0.0] - 2019-08-30
## Added
- New REST API method `get_towns_percentile_age_stats` in `/service/service_api.py`
- New service framework method `get_towns_percentile_age_stats` in `/service/service_framework.py`
- New return type `TownsPercentileAgeStats` in `/service/service_framework.py`
- New database framework method `get_towns_citizens_age_stats`
 and private method `_calculate_citizen_age` in `/service/database_framework.py`
 - New return type `TownsCitizensAgeStats` in `/service/database_framework.py`
- Tests for new REST API method (class `TestGetTownsPercentileAgeStats` in `/tests/test_service.py`)
- `calculate_citizen_age` method in `/tests/test_service.py`
- `numpy` to `requirements.txt`
- Documentation for new REST API method to README file
- Info about `numpy` module to README file

## Changed
- `BirthdayFmt` renamed to `BirthDateFmt`

## Removed
- Unnecessary `BirthdayFmt` variable from `/service/service_framework.py`


## [3.1.0] - 2019-08-30
## Added
- Custom Validator class to `/service/service_framework.py`
- `_prepare_citizens` private method to `/service/database_framework.py`

## Changed
- Data validation methods moved to Validator class
- Birthdate storing methods from `str` to `datetime.datetime`
- `get_citizen_relatives` moved to private method `_get_citizen_relatives`
- `get_import_citizens_birthdays` method from `/service/service_framework.py` reworked to operate with `datetime.datetime`
- Service tests reworked to operate with `datetime.datetime`

## [3.0.1] - 2019-08-30
## Changed
- Double quotes replaced with single quotes in string variables

## [3.0.0] - 2019-08-28
## Added
- `/tests/__init__` file for adding path to service modules to `sys.path`
- `NewType` variables for all types used in service modules
- Type hints to methods in `/service/service_api.py`, `/tests/test_service.py`, `/tests/conftest,py`
- Type hints to methods in `/service/database_framework.py` and `/service/service_framework` 
in places where they were forgotten
- Steps for removing `_id` field from data returned from database

## Changed
- Service architecture:
  - `conftest.py` and `test_service.py` moved to `/tests` folder
  - API module moved to `/service/service_api.py`
  - API requests processing logic moved to `/service/service_framework.py`
  - `db.py` module renamed to `database_framework.py` and moved to `/service/database_framework.py`
  - WSGI configuration file moved to `/service/wsgi.py`
- Documentation about service initialization in README file
- Documentation about service testing in README file

## [2.0.0] - 2019-08-28
## Added
- `get_citizens_with_relatives_dict` method to database framework
- `BirthdayFmt` variable to service
- New REST API method `get_birthdays`
- Test for new REST API method (class `TestGetBirthdays`)
- Documentation for new REST API method to README file
- `test_get_incorrect_import_citizens` test method

## Changed
- `TestGet` renamed to `TestGetCitizens`

## Removed
- Unnecessary sleep from tests
- `import time` from tests

## [1.1.1] - 2019-08-28
## Added
- `generate_10000_citizens_with_1000_relations` method to tests
- `test_patch_citizen_request_time` test method
- `test_get_citizens_request_time` test method

## Changed
- `test_big_import_request_time` test method structure
- `test_big_import_request_time` renamed to `test_import_request_time`
- `test_correct_citizens_patch` test method

## [1.1.0] - 2019-08-28
## Changed
- Request data validation algorithms

## Fixed
- `test_get_citizens`

## [1.0.6] - 2019-08-27
## Added
- Service test instructions to README file

## [1.0.5] - 2019-08-27
## Added
- Service start instructions to README file

## [1.0.4] - 2019-08-27
## Added
- Service setup instructions to README file
- qunicorn to `requirements.txt`

## [1.0.3] - 2019-08-27
## Added
- Table of contents to README file
- Additional info about service to README file
- Service control instructions to README file

## Fixed
- Typos in README file

## [1.0.2] - 2019-08-27
## Added
- Info about service to README file

### Changed
- Release format in CHANGELOG file

### Fixed
- Data of 1.0.1 release in CHANGELOG file

## [1.0.1] - 2019-08-27
### Added
- CHANGELOG file (`CHANGELOG.md`)

## [1.0.0] - 2019-08-26
### Added
- Base service program (`service.py`)
- DataBase framework to operate with MongoDB (`db.py`)
- Service tests (`test_service.py`)
- Test configuration file (`conftest.py`)
- Service WSGI application (`wsgi.py`)
- Service requirements file (`requirements.txt`)
- README file (`README.md`)


tests/__init__.py
import os
import sys

sys.path.append(os.path.join(os.path.dirname(__file__), os.pardir, "service"))


tests/conftest.py
from _pytest.config import argparsing


def pytest_addoption(parser: argparsing.Parser):
    parser.addoption('--service-address', action='store', default='127.0.0.1:5000')


db.py
from typing import Any, Dict, List, Optional

import pymongo


class DataBase:
    def __init__(self):
        self.client = pymongo.MongoClient('localhost', 27017)
        self.db = self.client['yandexbackend']

    def __del__(self):
        self.close()

    @property
    def imports(self) -> Optional[List[int]]:
        return sorted(int(import_name) for import_name in self.db.list_collection_names())

    def close(self):
        self.client.close()

    def insert_citizens_to_new_import(self, citizens: list):
        imports = self.imports
        new_import_id = imports[-1] + 1 if imports else 1
        new_import = self.db[f'{new_import_id}']
        new_import.insert_many(citizens)
        return new_import_id

    def get_all_import_citizens(self, import_id: int):
        import_ = self.db[f'{import_id}']
        citizens = list(import_.find({}))
        for citizen in citizens:
            citizen.pop('_id')
        return sorted(citizens, key=lambda c: c['citizen_id'])

    def get_citizens_with_relatives_dict(self, import_id: int):
        import_ = self.db[f'{import_id}']
        citizens = list(import_.find({'relatives': {'$not': {'$size': 0}}}))
        citizens_with_relatives_dict = {}
        for citizen in citizens:
            citizen.pop('_id')
        citizens = sorted(citizens, key=lambda c: len(c['relatives']))
        for citizen in citizens:
            citizens_with_relatives_dict[citizen['citizen_id']] = citizen
        return citizens_with_relatives_dict

    def get_citizen(self, import_id: int, citizen_id: int):
        import_ = self.db[f'{import_id}']
        citizen = import_.find_one({'citizen_id': citizen_id})
        return citizen

    def get_citizen_relatives(self, import_id: int, citizen_id: int):
        import_ = self.db[f'{import_id}']
        citizen = self.get_citizen(import_id, citizen_id)
        citizen_relatives = citizen['relatives']
        citizen_relatives = [import_.find_one({'citizen_id': related_id}) for related_id in citizen_relatives]
        return citizen_relatives

    def check_if_citizen_in_import(self, import_id: int, citizen_id: int):
        import_ = self.db[f'{import_id}']
        citizen = import_.find_one({'citizen_id': citizen_id})
        return True if citizen else False

    def update_citizen(self, import_id: int, citizen_id: int, citizen: Dict[str, Any]):
        import_ = self.db[f'{import_id}']
        import_.update_one({'citizen_id': citizen_id}, {"$set": citizen})

    def remove_citizen_from_old_relatives(self, import_id: int, citizen_id: int):
        citizen_relatives = self.get_citizen_relatives(import_id, citizen_id)
        for related_citizen in citizen_relatives:
            related_citizen_relatives = related_citizen['relatives']
            related_citizen_relatives.remove(citizen_id)
            self.update_citizen(import_id, related_citizen['citizen_id'], {'relatives': related_citizen_relatives})

    def add_citizen_to_new_relatives(self, import_id: int, citizen_id: int):
        citizen_relatives = self.get_citizen_relatives(import_id, citizen_id)
        for related_citizen in citizen_relatives:
            related_citizen_relatives = related_citizen['relatives']
            related_citizen_relatives.append(citizen_id)
            self.update_citizen(import_id, related_citizen['citizen_id'], {'relatives': related_citizen_relatives})


service.py
import collections
import copy
import datetime
import enum
import multiprocessing
import re
from typing import Any, Dict, List, Optional, Tuple, Union

import cerberus
import flask

import db

Counter = Dict[int, int]
CitizenDict = Dict[str, Any]
CitizensList = List[CitizenDict]
ValidatedCitizen = Dict[str, Union[int, str, List[int]]]
ValidatedCitizensList = List[ValidatedCitizen]
BirthdayFmt = '%d.%m.%Y'


app = flask.Flask(__name__)
data_base = db.DataBase()


class CitizenAttr(enum.Enum):
    ID = 'citizen_id'
    TOWN = 'town'
    STREET = 'street'
    BUILDING = 'building'
    APARTMENT = 'apartment'
    NAME = 'name'
    BIRTH_DATE = 'birth_date'
    GENDER = 'gender'
    RELATIVES = 'relatives'


def _validate_string(field, value, error):
    if value is None:
        error(field, 'Could not be NoneType')
    else:
        pattern = re.compile(r'[^\W_]')
        if not pattern.match(value):
            error(field, 'Must contain at least one letter or digit')


def _validate_date(field, value, error):
    try:
        date = datetime.datetime.strptime(value, BirthdayFmt).date()
    except (TypeError, ValueError):
        error(field, 'Wrong date format.')
    else:
        if date > datetime.datetime.utcnow().date():
            error(field, 'Birth date could not be more than current date.')


class CitizenValidator:
    CITIZEN_VALIDATION_SCHEMA = {
        'citizen_id': {'type': 'integer', 'min': 0, 'coerce': lambda v: None if isinstance(v, bool) else v},
        'town': {'type': 'string', 'minlength': 1, 'maxlength': 256, 'check_with': _validate_string},
        'street': {'type': 'string', 'minlength': 1, 'maxlength': 256, 'check_with': _validate_string},
        'building': {'type': 'string', 'minlength': 1, 'maxlength': 256, 'check_with': _validate_string},
        'apartment': {'type': 'integer', 'min': 0, 'coerce': lambda v: None if isinstance(v, bool) else v},
        'name': {'type': 'string', 'minlength': 1, 'maxlength': 256},
        'birth_date': {'check_with': _validate_date},
        'gender': {'type': 'string', 'allowed': ['male', 'female']},
        'relatives': {'type': 'list', 'schema': {'type': 'integer'}},
    }
    CITIZEN_VALIDATION_SCHEMA_WITHOUT_ID = copy.deepcopy(CITIZEN_VALIDATION_SCHEMA)
    VALIDATOR = cerberus.Validator(CITIZEN_VALIDATION_SCHEMA, require_all=True)

    @classmethod
    def validate_citizen(cls, citizen: Dict[str, Any]) -> Optional[Tuple[ValidatedCitizen, Counter, Counter]]:
        validated_citizen = cls.VALIDATOR.validated(citizen)
        if not validated_citizen:
            return None
        relatives = validated_citizen['relatives']
        if validated_citizen['citizen_id'] in relatives:
            return None
        return (
            validated_citizen,
            {validated_citizen['citizen_id']: len(relatives)} if relatives else {},
            {related_id: 1 for related_id in relatives},
        )


class CitizensValidator:
    @classmethod
    def validate_import_citizens(cls, citizens: CitizensList) -> Optional[ValidatedCitizensList]:
        validated_citizens_list = list()
        citizen_relatives = collections.Counter()
        related_citizens = collections.Counter()
        pool = multiprocessing.Pool(5)
        validated_citizens = [
            pool.apply_async(CitizenValidator.validate_citizen, args=(citizen,)) for citizen in citizens
        ]
        for res in validated_citizens:
            result = res.get()
            if not result:
                break
            validated_citizen, citizen_relatives_count, related_citizens_counts = result
            validated_citizens_list.append(validated_citizen)
            citizen_relatives.update(citizen_relatives_count)
            related_citizens.update(related_citizens_counts)
        pool.close()
        pool.join()
        if len(validated_citizens) != len(citizens):
            return None
        if not citizen_relatives == related_citizens:
            return None
        return validated_citizens_list


def get_import_citizens_birthdays(import_id: int):
    presents = {month: collections.defaultdict(int) for month in range(1, 13)}
    birthdays = {str(month): [] for month in range(1, 13)}
    citizens_with_relatives_dict = data_base.get_citizens_with_relatives_dict(import_id)
    citizen_ids = list(citizens_with_relatives_dict.keys())
    for citizen_id in citizen_ids:
        citizen = citizens_with_relatives_dict.get(citizen_id)
        if not citizen:
            continue
        citizen_relatives = citizen['relatives']
        citizen_birth_date = datetime.datetime.strptime(citizen['birth_date'], BirthdayFmt).date()
        for related_citizen_id in citizen_relatives:
            related_citizen = citizens_with_relatives_dict[related_citizen_id]
            related_citizen_birth_date = datetime.datetime.strptime(related_citizen['birth_date'], BirthdayFmt).date()
            related_citizen_relatives = related_citizen['relatives']
            presents[related_citizen_birth_date.month][citizen_id] += 1
            presents[citizen_birth_date.month][related_citizen_id] += 1
            related_citizen_relatives.remove(citizen_id)
            if not related_citizen_relatives:
                citizens_with_relatives_dict.pop(related_citizen_id)
    for month, citizen_presents in presents.items():
        for citizen_id, presents in citizen_presents.items():
            birthdays[str(month)].append({'citizen_id': citizen_id, 'presents': presents})
        birthdays[str(month)] = sorted(birthdays[str(month)], key=lambda c: c['citizen_id'])
    return birthdays


@app.route('/imports', methods=['POST'])
def import_citizens():
    data = flask.request.json
    if not isinstance(data, dict):
        return flask.make_response('Bad Request', 400)
    schema = {'citizens': {'type': 'list'}}
    validator = cerberus.Validator(schema, require_all=True)
    if not validator.validate(data):
        return flask.make_response('Bad Request', 400)
    validated_citizens = CitizensValidator.validate_import_citizens(data['citizens'])
    if not validated_citizens:
        return flask.make_response('Bad Request', 400)
    new_import_id = data_base.insert_citizens_to_new_import(validated_citizens)
    resp_json = {'data': {'import_id': new_import_id}}
    return flask.make_response(flask.jsonify(resp_json), 201)


@app.route('/imports/<int:import_id>/citizens/<int:citizen_id>', methods=['PATCH'])
def patch_citizen(import_id: int, citizen_id: int):
    if import_id not in data_base.imports:
        return flask.make_response('Bad Request', 400)
    citizen = data_base.get_citizen(import_id, citizen_id)
    if not citizen:
        return flask.make_response('Bad Request', 400)
    data = flask.request.json
    if not data or not isinstance(data, dict):
        return flask.make_response('Bad Request', 400)
    if 'citizen_id' in data:
        return flask.make_response('Bad Request', 400)
    new_citizen = copy.deepcopy(citizen)
    new_citizen.pop('_id')
    new_citizen.update(data)
    validation_result = CitizenValidator.validate_citizen(new_citizen)
    if not validation_result:
        return flask.make_response('Bad Request', 400)
    validated_citizen, *a = validation_result
    relatives = validated_citizen.get('relatives')
    if not all(data_base.check_if_citizen_in_import(import_id, relative_id) for relative_id in relatives):
        return flask.make_response('Bad Request', 400)
    if any(relative_id == citizen_id for relative_id in relatives):
        return flask.make_response('Bad Request', 400)
    validated_citizen.pop('citizen_id')
    old_relatives = citizen['relatives']
    new_relatives = validated_citizen['relatives']
    if old_relatives != new_relatives:
        data_base.remove_citizen_from_old_relatives(import_id, citizen_id)
        data_base.update_citizen(import_id, citizen_id, validated_citizen)
        data_base.add_citizen_to_new_relatives(import_id, citizen_id)
    else:
        data_base.update_citizen(import_id, citizen_id, validated_citizen)
    updated_citizen = data_base.get_citizen(import_id, citizen_id)
    updated_citizen.pop('_id')
    resp_json = {'data': updated_citizen}
    return flask.make_response(flask.jsonify(resp_json), 200)


@app.route('/imports/<int:import_id>/citizens', methods=['GET'])
def get_citizens(import_id: int):
    if import_id not in data_base.imports:
        return flask.make_response('Bad Request', 400)
    resp_json = {'data': data_base.get_all_import_citizens(import_id)}
    return flask.make_response(flask.jsonify(resp_json), 200)


@app.route('/imports/<int:import_id>/citizens/birthdays', methods=['GET'])
def get_birthdays(import_id: int):
    if import_id not in data_base.imports:
        return flask.make_response('Bad Request', 400)
    resp_json = {'data': get_import_citizens_birthdays(import_id)}
    return flask.make_response(flask.jsonify(resp_json), 200)
